src/checkers/inference/util/JaifBuilder.java

```diff
@@ -256,22 +257,59 @@ private void buildClassEntries() {
         }
     }
 
-    private boolean isGenericOrArrayEntry(String pathString) {
-        return     pathString.contains("ParameterizedType.typeArgument")
-                || pathString.contains("ArrayType.type");
-    }
+    /**
+     * @param astRecord
+     * @return true if the given AST path represents a main modifier of a local variable
+     * An AST Path represents a main modifier of a local variable should have pattern like
+     * 1) ..., Block.statement #, ..., Variable.type
+     * 2) ..., Block.statement #, ..., Variable.type, ParameterizedType.type
+     * reference: Local Variable Declaration Statements in JLS8
+     * https://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.4
+     */
+    protected boolean isMainModOfLocalVar(ASTPath astPath) {
+        Iterator<ASTEntry> iterator = astPath.iterator();
 
-    private boolean isReturnOrParameterEntry(String pathString) {
-        return     pathString.startsWith("Method.type")
-                || pathString.startsWith("Method.parameter -1")
-                || pathString.startsWith("Variable.initializer")
-                || pathString.matches(paramPathRegex);
-    }
+        // first determine whether this astPath is a block statement
+        while (iterator.hasNext()) {
+            if (isEntryIs(Tree.Kind.BLOCK, ASTPath.STATEMENT, iterator.next())) {
+                break;
+            }
+        }
+
+        if (!iterator.hasNext()) {
+            // this astPath either does not has Block.statement, or end up with Block.statement
+            // in both cases it doesn't represent a main modifier of a local variable
+            return false;
+        }
+
+        // next get the last two entry of this AST Path
+        ASTEntry prevEntry = null;
+        ASTEntry leafEntry = null;
+        while (iterator.hasNext()) {
+            leafEntry = iterator.next();
+            if (!iterator.hasNext()) {
+                break;
+            }
+            prevEntry = leafEntry;
+        }
+
+        assert leafEntry != null;
 
-    private boolean isMethodEntry(ASTRecord record) {
-        return record.methodName != null && record.varName == null;
+        if (isEntryIs(Tree.Kind.VARIABLE, ASTPath.TYPE, leafEntry)) {
+            // the first kind of AST path of main modifier of local variable
+            return true;
+        } else if (prevEntry != null && isEntryIs(Tree.Kind.VARIABLE, ASTPath.TYPE, prevEntry) &&
+            isEntryIs(Tree.Kind.PARAMETERIZED_TYPE, ASTPath.TYPE, leafEntry)) {
+            // the second kind
+            return true;
+        }
+
+        return false;
     }
 
+    protected static boolean isEntryIs(Tree.Kind kind, String childSelector, ASTEntry entry) {
```

Comment:
Already add javadoc for this method, and also change the method name to `isEntry`